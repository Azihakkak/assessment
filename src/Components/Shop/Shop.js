import React, { Component } from "react";

import CartView from "../CartView/CartView";

const items = [
  { name: "Shower gel", price: 49.99 },
  { name: "Deodorant", price: 99.99, id: 2 },
];

export default class Shop extends Component {
  constructor() {
    super();
    this.state = {
      selectedItem: "Shower gel",
      itemQty: 1,
      cart: {},
    };
  }

  handleChange = (event) => {
    const { value, name } = event.target;
    this.setState({
      [name]: value,
    });
  };

  handleClick = () => {
    this.setState({
      itemQty: this.state.itemQty + 1,
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const selectedItem = this.state.selectedItem;
    const qty = this.state.itemQty;
    let cart = this.state.cart;

    const newQuantity = cart[selectedItem]
      ? +cart[selectedItem].quantity + +qty
      : qty;
    const item = items.find((item) => item.name === selectedItem);

    cart[selectedItem] = {
      quantity: newQuantity,
      price: item.price,
    };

    this.setState({ ...this.state, cart });
  };

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <label>
            Select your item
            <select
              value={this.state.selectedItem}
              onChange={this.handleChange}
              name="selectedItem"
            >
              <option value="Shower gel">Shower gel</option>
              <option value="Deodorant">Deodorant</option>
            </select>
          </label>
          <div>
            <input
              name="itemQty"
              type="number"
              value={this.state.itemQty}
              onChange={this.handleChange}
            />
          </div>
          <input type="submit" value="Add to cart" />
        </form>
        <CartView cart={this.state.cart} />
        {/* cart.item name cart.quantity cart.qty * price  */}
      </div>
    );
  }
}

// user adds 5 shower gels to the cart
// shopping cart should show 5 shower gels with unit price
// total price should be 249.95
