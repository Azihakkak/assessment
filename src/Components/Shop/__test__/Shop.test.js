import React from "react";
import ReactDOM from "react-dom";
import Shop from "../Shop";
import { render, fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

it("renders correctly", () => {
  const { queryByTestId } = render(<Shop />);

  expect(queryByTestId("item-selection")).toBeTruthy();
  expect(queryByTestId("qty")).toBeTruthy();
});

describe("select value", () => {
  it("updates on change", () => {
    const { queryByTestId } = render(<Shop />);

    const selectInput = queryByTestId("item-selection");

    fireEvent.change(selectInput, { target: { value: "Shower gel" } });

    expect(selectInput.value).toBe("Shower gel");
  });
});

describe("select amount", () => {
  it("add quantity", () => {
    const { queryByTestId } = render(<Shop />);

    const inputQty = queryByTestId("qty");

    fireEvent.change(inputQty, { target: { value: "5" } });

    expect(inputQty.value).toBe("5");
  });
});

describe("when user adds 5 Shower gels to the cart ", () => {
  it("should return showerGelQty: 5, unitPrice: 49.99, priceExcludingTax: 249.95 ", () => {
    const { queryByTestId, getByText } = render(<Shop />);

    const selectInput = queryByTestId("item-selection");

    fireEvent.change(selectInput, { target: { value: "Shower gel" } });

    const inputQty = queryByTestId("qty");

    fireEvent.change(inputQty, { target: { value: "5" } });

    const button = queryByTestId("button");

    fireEvent.click(button);

    expect(getByText("Total Price: $249.95")).toBeInTheDocument();
  });
});

describe("when user adds 5 Shower gels and again adds 3 more shower gels to the cart ", () => {
  it("should return showerGelQty: 8, unitPrice: 49.99, priceExcludingTax: 399.92 ", () => {
    const { queryByTestId, getByText } = render(<Shop />);

    const selectInput = queryByTestId("item-selection");

    fireEvent.change(selectInput, { target: { value: "Shower gel" } });

    const inputQty = queryByTestId("qty");

    fireEvent.change(inputQty, { target: { value: "5" } });

    const selectInput1 = queryByTestId("item-selection");

    fireEvent.change(selectInput1, { target: { value: "Shower gel" } });

    const inputQty2 = queryByTestId("qty");

    fireEvent.change(inputQty2, { target: { value: "3" } });

    const button = queryByTestId("button");

    fireEvent.click(button);

    expect(getByText("Total Price: $399.92")).toBeInTheDocument();
  });
});
