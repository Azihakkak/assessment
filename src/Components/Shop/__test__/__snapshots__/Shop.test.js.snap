// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
      Select your item
      <select
        data-testid="item-selection"
        name="selectedItem"
        onChange={[Function]}
        value="Shower gel"
      >
        <option
          value="Shower gel"
        >
          Shower gel
        </option>
        <option
          value="Deodorant"
        >
          Deodorant
        </option>
      </select>
    </label>
    <div>
      <input
        data-testid="qty"
        name="itemQty"
        onChange={[Function]}
        type="number"
        value={1}
      />
    </div>
    <input
      data-testid="button"
      type="submit"
      value="Add to cart"
    />
  </form>
  <div
    data-testid="cart-view"
  >
    <div>
      <table>
        <tbody />
      </table>
      <h3>
        Total Tax: $
        0.01
      </h3>
      <h3>
        Total Price: $
        0.01
      </h3>
      <h3>
        Total Price IncludingTax: $
        0.02
      </h3>
    </div>
  </div>
</div>
`;
